File: packages/python/codex/src/anosys_sdk_codex/mapper.py
"""
anosys_sdk_codex.mapper — Schema mapper for Codex CLI hook turns.
Maps OpenAI Codex rollout session turns and tool calls into the AnoSys ingestion format.
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

INTEGRATION_VERSION = "0.1.0"
OS_USER = os.environ.get("USERNAME") or os.environ.get("USER") or "unknown"
REDACTION = os.environ.get("REDACTION", "false").lower() == "true"

OPENAI_PRICING: Dict[str, Dict[str, float]] = {
    "gpt-4o": {
        "input": 2.50 / 1e6,
        "output": 10.00 / 1e6,
        "cached": 1.25 / 1e6,
    },
    "gpt-4o-mini": {
        "input": 0.15 / 1e6,
        "output": 0.60 / 1e6,
        "cached": 0.075 / 1e6,
    },
    "o1": {
        "input": 15.00 / 1e6,
        "output": 60.00 / 1e6,
        "cached": 7.50 / 1e6,
    },
    "o1-mini": {
        "input": 1.10 / 1e6,
        "output": 4.40 / 1e6,
        "cached": 0.55 / 1e6,
    },
    "o3": {
        "input": 15.00 / 1e6,
        "output": 60.00 / 1e6,
        "cached": 7.50 / 1e6,
    },
    "o3-mini": {
        "input": 1.10 / 1e6,
        "output": 4.40 / 1e6,
        "cached": 0.55 / 1e6,
    },
    "default": {
        "input": 2.50 / 1e6,
        "output": 10.00 / 1e6,
        "cached": 1.25 / 1e6,
    },
}


def calculate_cost(
    model: str,
    input_tokens: int,
    output_tokens: int,
    cached_tokens: int = 0,
) -> float:
    rates = OPENAI_PRICING.get("default", {"input": 2.5e-6, "output": 1.0e-5, "cached": 1.25e-6})
    for m_key, r in OPENAI_PRICING.items():
        if m_key in model.lower():
            rates = r
            break

    billable_input = max(0, input_tokens - cached_tokens)
    cost = (billable_input * rates["input"]) + (cached_tokens * rates["cached"]) + (output_tokens * rates["output"])
    return round(cost, 6)


def extract_commands_and_paths(tool_calls: List[Dict[str, Any]]) -> tuple[List[str], List[str]]:
    commands: List[str] = []
    written_paths: List[str] = []

    for call in tool_calls:
        tool = str(call.get("tool", "")).lower()
        args = call.get("args") or ""

        if tool in ("shell", "bash", "exec", "terminal"):
            cmd = ""
            if isinstance(args, dict):
                cmd = args.get("command") or args.get("cmd") or ""
            elif isinstance(args, str):
                try:
                    parsed = json.loads(args)
                    if isinstance(parsed, dict):
                        cmd = parsed.get("command") or parsed.get("cmd") or ""
                except Exception:
                    cmd = args
            if cmd:
                commands.append(cmd.strip())

        if tool in ("apply_patch", "write_to_file", "replace_file_content", "edit", "save"):
            fp = ""
            if isinstance(args, dict):
                fp = args.get("path") or args.get("target_file") or args.get("file") or ""
            elif isinstance(args, str):
                try:
                    parsed = json.loads(args)
                    if isinstance(parsed, dict):
                        fp = parsed.get("path") or parsed.get("target_file") or parsed.get("file") or ""
                except Exception:
                    fp = args
            if fp:
                written_paths.append(fp.strip())

    return commands, written_paths


def transform_codex_turn(
    turn: Dict[str, Any],
    session_id: str,
    turn_id: str,
    incremental_tokens: Optional[Dict[str, float]] = None,
    redact: bool | None = None,
) -> Dict[str, Any]:
    do_redact = REDACTION if redact is None else redact

    model = turn.get("model") or "unknown"
    provider = turn.get("model_provider") or "openai"
    cwd = turn.get("cwd") or ""
    project = Path(cwd).name if cwd else None

    user_prompt = turn.get("user_prompt") or ""
    assistant_text = turn.get("assistant_output") or ""

    if do_redact:
        user_prompt = "[REDACTED]"
        assistant_text = "[REDACTED]"

    tokens = turn.get("tokens") or {}
    input_tokens = int(tokens.get("input_tokens", 0))
    output_tokens = int(tokens.get("output_tokens", 0))
    total_tokens = int(tokens.get("total_tokens", input_tokens + output_tokens))
    cached_tokens = int(tokens.get("cached_input_tokens", 0))
    cache_write_tokens = int(tokens.get("cache_write_input_tokens", 0))
    reasoning_tokens = int(tokens.get("reasoning_output_tokens", 0))

    cost_est = calculate_cost(model, input_tokens, output_tokens, cached_tokens)

    inc = incremental_tokens or {
        "input": float(input_tokens),
        "output": float(output_tokens),
        "total": float(total_tokens),
        "cache_read": float(cached_tokens),
        "cache_creation": float(cache_write_tokens),
        "cost": float(cost_est),
    }

    tool_calls = turn.get("tool_calls") or []
    commands, written_paths = extract_commands_and_paths(tool_calls)

    tool_calls_clean = []
    for tc in tool_calls:
        clean_entry = dict(tc)
        if do_redact:
            clean_entry["args"] = "[REDACTED]"
            clean_entry["output"] = "[REDACTED]"
        tool_calls_clean.append(clean_entry)

    tool_durations = [
        int(tc.get("end_ts", 0)) - int(tc.get("start_ts", 0))
        for tc in tool_calls
        if tc.get("end_ts") and tc.get("start_ts") and tc.get("end_ts") >= tc.get("start_ts")
    ]
    tool_duration_ms = sum(tool_durations) if tool_durations else None

    start_ts = turn.get("turn_start_ms") or int(time.time() * 1000)
    duration_ms = turn.get("duration_ms")

    payload: Dict[str, Any] = {
        "sessionId": session_id,
        "uuid": turn_id,
        "eventId": turn_id,
        "timestamp": start_ts,
        "userPrompt": user_prompt,
        "assistantText": assistant_text,
        "model": model,
        "model_provider": provider,
        "cwd": cwd or None,
        "project": project,
        "permissionMode": turn.get("permission_mode") or None,
        "sandbox_mode": turn.get("sandbox_mode") or None,
        "duration_ms": duration_ms,
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "total_tokens": total_tokens,
        "cache_read": cached_tokens,
        "cache_creation": cache_write_tokens,
        "reasoning_tokens": reasoning_tokens,
        "cost_estimate": cost_est,
        "incremental_input": inc.get("input", input_tokens),
        "incremental_output": inc.get("output", output_tokens),
        "incremental_total": inc.get("total", total_tokens),
        "incremental_cost": inc.get("cost", cost_est),
        "incremental_cache_read": inc.get("cache_read", cached_tokens),
        "incremental_cache_creation": inc.get("cache_creation", cache_write_tokens),
        "tool_count": len(tool_calls),
        "tool_duration_ms": tool_duration_ms,
        "tools_used": [tc.get("tool") for tc in tool_calls if tc.get("tool")],
        "commands": commands if commands else None,
        "written_paths": written_paths if written_paths else None,
        "has_thinking": reasoning_tokens > 0,
        "integration_version": INTEGRATION_VERSION,
        "os_user": OS_USER,
        "cvs199": json.dumps(tool_calls_clean, separators=(",", ":")),
        "cvs200": "CodexHook",
    }

    return {k: v for k, v in payload.items() if v is not None}

File: packages/js/codex/src/mapper.js
'use strict';

const path = require('path');
const os = require('os');

const INTEGRATION_VERSION = require('../package.json').version || '0.1.0';
const OS_USER = os.userInfo().username || 'unknown';

const OPENAI_PRICING = {
  'gpt-4o': { input: 2.50 / 1e6, output: 10.00 / 1e6, cached: 1.25 / 1e6 },
  'gpt-4o-mini': { input: 0.15 / 1e6, output: 0.60 / 1e6, cached: 0.075 / 1e6 },
  'o1': { input: 15.00 / 1e6, output: 60.00 / 1e6, cached: 7.50 / 1e6 },
  'o1-mini': { input: 1.10 / 1e6, output: 4.40 / 1e6, cached: 0.55 / 1e6 },
  'o3': { input: 15.00 / 1e6, output: 60.00 / 1e6, cached: 7.50 / 1e6 },
  'o3-mini': { input: 1.10 / 1e6, output: 4.40 / 1e6, cached: 0.55 / 1e6 },
  'default': { input: 2.50 / 1e6, output: 10.00 / 1e6, cached: 1.25 / 1e6 }
};

function calculateCost(model, inputTokens, outputTokens, cachedTokens = 0) {
  const m = (model || '').toLowerCase();
  let rates = OPENAI_PRICING.default;
  for (const [k, r] of Object.entries(OPENAI_PRICING)) {
    if (m.includes(k)) {
      rates = r;
      break;
    }
  }
  const billableInput = Math.max(0, inputTokens - cachedTokens);
  const cost = (billableInput * rates.input) + (cachedTokens * rates.cached) + (outputTokens * rates.output);
  return Number(cost.toFixed(6));
}

function extractCommandsAndPaths(toolCalls) {
  const commands = [];
  const writtenPaths = [];

  for (const call of (toolCalls || [])) {
    const tool = String(call.tool || '').toLowerCase();
    const args = call.args || '';

    if (['shell', 'bash', 'exec', 'terminal'].includes(tool)) {
      let cmd = '';
      if (typeof args === 'object' && args !== null) {
        cmd = args.command || args.cmd || '';
      } else if (typeof args === 'string') {
        try {
          const parsed = JSON.parse(args);
          if (typeof parsed === 'object' && parsed !== null) {
            cmd = parsed.command || parsed.cmd || '';
          }
        } catch (_) {
          cmd = args;
        }
      }
      if (cmd && typeof cmd === 'string') commands.push(cmd.trim());
    }

    if (['apply_patch', 'write_to_file', 'replace_file_content', 'edit', 'save'].includes(tool)) {
      let fp = '';
      if (typeof args === 'object' && args !== null) {
        fp = args.path || args.target_file || args.file || '';
      } else if (typeof args === 'string') {
        try {
          const parsed = JSON.parse(args);
          if (typeof parsed === 'object' && parsed !== null) {
            fp = parsed.path || parsed.target_file || parsed.file || '';
          }
        } catch (_) {
          fp = args;
        }
      }
      if (fp && typeof fp === 'string') writtenPaths.push(fp.trim());
    }
  }

  return { commands, writtenPaths };
}

function transformCodexTurn(turn, sessionId, turnId, incrementalTokens = null, redact = null) {
  const envRedaction = (process.env.REDACTION || 'false').toLowerCase() === 'true';
  const doRedact = redact !== null ? redact : envRedaction;

  const model = turn.model || 'unknown';
  const provider = turn.model_provider || 'openai';
  const cwd = turn.cwd || '';
  const project = cwd ? path.basename(cwd) : null;

  let userPrompt = turn.user_prompt || '';
  let assistantText = turn.assistant_output || '';

  if (doRedact) {
    userPrompt = '[REDACTED]';
    assistantText = '[REDACTED]';
  }

  const tokens = turn.tokens || {};
  const inputTokens = Number(tokens.input_tokens || 0);
  const outputTokens = Number(tokens.output_tokens || 0);
  const totalTokens = Number(tokens.total_tokens || (inputTokens + outputTokens));
  const cachedTokens = Number(tokens.cached_input_tokens || 0);
  const cacheWriteTokens = Number(tokens.cache_write_input_tokens || 0);
  const reasoningTokens = Number(tokens.reasoning_output_tokens || 0);

  const costEst = calculateCost(model, inputTokens, outputTokens, cachedTokens);

  const inc = incrementalTokens || {
    input: inputTokens,
    output: outputTokens,
    total: totalTokens,
    cache_read: cachedTokens,
    cache_creation: cacheWriteTokens,
    cost: costEst
  };

  const toolCalls = turn.tool_calls || [];
  const { commands, writtenPaths } = extractCommandsAndPaths(toolCalls);

  const cleanToolCalls = toolCalls.map(tc => {
    const c = Object.assign({}, tc);
    if (doRedact) {
      c.args = '[REDACTED]';
      c.output = '[REDACTED]';
    }
    return c;
  });

  const toolDurations = toolCalls
    .map(tc => (tc.end_ts && tc.start_ts && tc.end_ts >= tc.start_ts ? (tc.end_ts - tc.start_ts) : 0))
    .filter(d => d > 0);
  const toolDurationMs = toolDurations.length ? toolDurations.reduce((a, b) => a + b, 0) : null;

  const startTs = turn.turn_start_ms || Date.now();
  const durationMs = turn.duration_ms !== undefined ? turn.duration_ms : null;

  const payload = {
    sessionId: sessionId,
    uuid: turnId,
    eventId: turnId,
    timestamp: startTs,
    userPrompt: userPrompt,
    assistantText: assistantText,
    model: model,
    model_provider: provider,
    cwd: cwd || null,
    project: project,
    permissionMode: turn.permission_mode || null,
    sandbox_mode: turn.sandbox_mode || null,
    duration_ms: durationMs,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    total_tokens: totalTokens,
    cache_read: cachedTokens,
    cache_creation: cacheWriteTokens,
    reasoning_tokens: reasoningTokens,
    cost_estimate: costEst,
    incremental_input: inc.input !== undefined ? inc.input : inputTokens,
    incremental_output: inc.output !== undefined ? inc.output : outputTokens,
    incremental_total: inc.total !== undefined ? inc.total : totalTokens,
    incremental_cost: inc.cost !== undefined ? inc.cost : costEst,
    incremental_cache_read: inc.cache_read !== undefined ? inc.cache_read : cachedTokens,
    incremental_cache_creation: inc.cache_creation !== undefined ? inc.cache_creation : cacheWriteTokens,
    tool_count: toolCalls.length,
    tool_duration_ms: toolDurationMs,
    tools_used: toolCalls.map(tc => tc.tool).filter(Boolean),
    commands: commands.length ? commands : null,
    written_paths: writtenPaths.length ? writtenPaths : null,
    has_thinking: reasoningTokens > 0,
    integration_version: INTEGRATION_VERSION,
    os_user: OS_USER,
    cvs199: JSON.stringify(cleanToolCalls),
    cvs200: 'CodexHook'
  };

  // Remove null/undefined properties
  const cleanPayload = {};
  for (const [k, v] of Object.entries(payload)) {
    if (v !== null && v !== undefined) {
      cleanPayload[k] = v;
    }
  }
  return cleanPayload;
}

module.exports = {
  calculateCost,
  extractCommandsAndPaths,
  transformCodexTurn
};
